[TS migration] Migrate Signin to TypeScript

fedirjh opened the pull request:
<!-- If necessary, assign reviewers that know the area or changes well. Feel free to tag any additional reviewers you see fit. -->

### Details

### Fixed Issues

$ https://github.com/Expensify/App/issues/25224
PROPOSAL: https://github.com/Expensify/App/issues/25224#issuecomment-1911977494


### Tests

Test 1: SAML enabled but not required

1. Create an account with a validated domain on expensify.com
2. Configure your account to have **SAML enabled but not required**
3. Enter the email for your account in New Expensify
4. Confirm you see the option to choose to sign in with SAML or with a magic code
5. Choose the Go Back option and confirm you're dropped back on the sign-in page with your email pre-filled
6. Press Continue
7. Choose the Use Magic Code option and confirm you get a magic code sent to your email
8. Press Go Back and confirm you're back on the sign-in page with your email pre-filled
9. Press Continue
10. Choose Use Single Sign On option
11. Confirm you briefly see an interstitial page about launching your SSO provider's login portal

Test 2: SAML enabled and required

1. Create an account with a validated domain on expensify.com
2. Configure your account to have **SAML enabled and required**
3. Enter the email for your account in New Expensify
4. Press Continue
5. Confirm you're automatically redirected to your SSO providers login page (with the interstitial page being briefly shown)


Test 3: Login / Signup

1. Open App and navigate to the sign in page.
2. Enter your email address and press Continue.
3. Verify Enter magic code screen appears
4. Enter wrong magic code
5. Verify an error message appears.
6. Click Go back and confirm you're dropped back on the sign-in page with your email pre-filled
7. Click continue and enter the correct magic code
8. Verify you're signed in successfully.

Test 4: Email delivery failure

1. Login into your account on oldDot and add the email delivery failed NVP by pasting the following in the JS console: `NVP.set('expensify_emailDeliveryFailure', {'lastSMTPErrorDatetime': moment.now().addDays(1).toString()});`
2. In an incognito window, try to log into your account in newDot
3. Confirm you see the email delivery failure screen.
4. Click Back and confirm it takes you back to the login view to enter an email/sms number
5. In your original window run `NVP.set('expensify_emailDeliveryFailure', null);` in the JS console to unblock your account
6. Click continue and confirm the magic code screen is displayed.

Test 5: Web Only

1. Open the Login page.
2. Open a new Login page in a new window or tab.
3. Check if `Another Login Page is Open` view is shown in the first window/tab.

Test 6: Footer Links

1. Open App and navigate to the sign in page.
2. Scroll to the footer 
3. Verify links works as expected.

- [x] Verify that the design of the sign in page matches the production version.
- [x] Verify that no errors appear in the JS console

### Offline tests

1. Navigate to the login page for New Expensify and enter your email
4. Disable your wifi or disable your network connections
5. Confirm you see the offline indicator

### QA Steps

Same as the Tests section above + :

1. Go to sign in page
3. Click the Google/Apple sign-in buttons
4. Verify that buttons works as expected

- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/Expensify/App/assets/36869046/c063bc67-b0de-4e12-8718-aeac852ae368

https://github.com/Expensify/App/assets/36869046/6ac158cd-e028-494d-bdbe-6987c5657c71

https://github.com/Expensify/App/assets/36869046/89b495d1-ff6a-4c72-a5b6-a0c05baaacb7

<img width="458" alt="Screenshot 2024-02-07 at 4 40 40 PM" src="https://github.com/Expensify/App/assets/36869046/e748b3bb-2112-42bf-8fb6-f8a118e5650c">

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/Expensify/App/assets/36869046/e97ca8ed-9238-4a13-bc7e-57c204ffb905


https://github.com/Expensify/App/assets/36869046/250fb5db-9f19-4055-a033-4c9fed5dcf56


https://github.com/Expensify/App/assets/36869046/bcbec847-2608-4ae0-bc7e-d26ad1836902

<img width="436" alt="Screenshot 2024-02-07 at 4 30 42 PM" src="https://github.com/Expensify/App/assets/36869046/8bbea033-dd13-4f7c-b225-c6d685ac2208">

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/Expensify/App/assets/36869046/58993279-2657-43b4-9313-781eb8612b37

https://github.com/Expensify/App/assets/36869046/7f2efd83-1c05-494d-9512-2e310c6b26bb

<img width="479" alt="Screenshot 2024-02-07 at 3 56 47 PM" src="https://github.com/Expensify/App/assets/36869046/88e306de-f1dc-4e73-b587-90533a95126d">

https://github.com/Expensify/App/assets/36869046/6bdac6f4-6431-4353-8729-8562be156a77

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/Expensify/App/assets/36869046/6177034b-9996-430c-be82-fc00fa4438c0

https://github.com/Expensify/App/assets/36869046/df56bd42-85cd-488d-81b3-99a62146ee0f

https://github.com/Expensify/App/assets/36869046/5d9c2668-48c0-4a30-a48c-13b0c164c2bc

<img width="479" alt="Screenshot 2024-02-07 at 4 09 45 PM" src="https://github.com/Expensify/App/assets/36869046/11cc69a8-300b-4397-9d49-1f35c4e3b1fd">

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/App/assets/36869046/4cb6af2e-bfad-4b86-b194-e22962060a49

https://github.com/Expensify/App/assets/36869046/09b84253-b921-40b3-8786-7273dc7ff090

https://github.com/Expensify/App/assets/36869046/2dc41ddf-6d7f-47ca-9c7f-aa8402c81a33

<img width="1440" alt="Screenshot 2024-02-07 at 3 13 51 PM" src="https://github.com/Expensify/App/assets/36869046/c448ef36-0245-4f1b-8f6a-2f12500c4a22">

https://github.com/Expensify/App/assets/36869046/c1a6cdca-01c6-4c5e-85d9-76a7c7e5c368

https://github.com/Expensify/App/assets/36869046/91e7e551-49dd-4c66-b232-839b2b72d59c

</details>

<details>
<summary>MacOS: Desktop</summary>
<img width="1236" alt="Screenshot 2024-02-07 at 3 23 43 PM" src="https://github.com/Expensify/App/assets/36869046/c1ba1d2d-a4e6-42b7-895c-5ae714cebc1d">

https://github.com/Expensify/App/assets/36869046/b6fc77b2-5ff2-4aad-97d0-1c7069bfd12c


https://github.com/Expensify/App/assets/36869046/796851de-da84-48cb-aa04-1811ff978a94


https://github.com/Expensify/App/assets/36869046/0e9a231b-4906-4bc6-a3de-a09791a537c1

</details>

fedirjh committed:
[Refactor] Convert Socials.js to TypeScript

fedirjh committed:
[Refactor] Convert Terms.js to TypeScript

fedirjh committed:
[Refactor] Convert Licenses.js to TypeScript

fedirjh committed:
[Refactor] Convert SignInPageHero.js to TypeScript

fedirjh committed:
[Refactor] Convert SignInPageHero.js to TypeScript

fedirjh committed:
[Refactor] Convert SignInModal.js to TypeScript

fedirjh committed:
[Refactor] Convert DesktopRedirectPage.js to TypeScript

fedirjh committed:
[Refactor] Convert DesktopRedirectPage.js to TypeScript

fedirjh committed:
[Refactor] Convert DesktopSignInRedirectPage to TypeScript

fedirjh committed:
[Refactor] Convert DesktopSignInRedirectPage to TypeScript

fedirjh committed:
[Refactor] Convert SAMLSignInPage to TypeScript

fedirjh committed:
Adjust semicolon usage and import organization

fedirjh committed:
Adjust semicolon usage and import organization

fedirjh committed:
[Refactor] Convert SignIn page styles to TypeScript.

fedirjh committed:
[Refactor] Convert ChangeExpensifyLoginLink to TypeScript.

fedirjh committed:
[Refactor] Convert ChangeExpensifyLoginLink to TypeScript.

fedirjh committed:
[Refactor] Convert ChangeExpensifyLoginLink to TypeScript.

fedirjh committed:
[Refactor] Convert ChangeExpensifyLoginLink to TypeScript.

fedirjh committed:
[Refactor] Convert SignInHeroImage to TypeScript.

fedirjh committed:
[Refactor] Convert SignInHeroCopy to TypeScript.

fedirjh committed:
[Refactor] Convert EmailDeliveryFailurePage to TypeScript.

fedirjh committed:
[Refactor] Convert UnlinkLoginForm to TypeScript.

fedirjh committed:
[Chore] Add comments

fedirjh committed:
[Refactor] Convert ChooseSSOOrMagicCode to TypeScript.

fedirjh committed:
[Refactor] Convert ThirdPartySignInPage to TypeScript.

fedirjh committed:
[Refactor] Convert BackgroundImage to TypeScript.

fedirjh committed:
[Chore] The TextLink component props definition was simplified by removing redundant props types and improving optionality checks.

fedirjh committed:
[Chore] The TextLink component props definition was simplified by removing redundant props types and improving optionality checks.

fedirjh committed:
[Refactor] convert to TypeScript

fedirjh committed:
[Refactor] convert to TypeScript

fedirjh committed:
[Refactor] convert ValidateCodeForm to TypeScript

fedirjh committed:
[Refactor] convert ValidateCodeForm to TypeScript

fedirjh committed:
[Refactor] convert LoginForm to TypeScript

fedirjh committed:
[Refactor] convert SignInPageContent to TypeScript

fedirjh committed:
[Refactor] convert SignInPageLayout to TypeScript

fedirjh committed:
[Refactor] convert SignInPageLayout to TypeScript

fedirjh committed:
Revert "[Refactor] Convert ChangeExpensifyLoginLink to TypeScript."

This reverts commit 9ea1cc2cf5777e5239181728934454f0a8762eb3.

fedirjh committed:
Revert "[Refactor] Convert ChangeExpensifyLoginLink to TypeScript."

This reverts commit 9ea1cc2cf5777e5239181728934454f0a8762eb3.

fedirjh committed:
Revert "[Refactor] Convert ChangeExpensifyLoginLink to TypeScript."

This reverts commit 8523026f48f50a9e0618fa6fb0598817c86d7460.

fedirjh committed:
Revert "[Refactor] Convert ChangeExpensifyLoginLink to TypeScript."

This reverts commit 8523026f48f50a9e0618fa6fb0598817c86d7460.

fedirjh committed:
Revert "[Refactor] convert ValidateCodeForm to TypeScript"

This reverts commit b782ad6b415d3d3f211a7005d275f1ddf4c903ab.

fedirjh committed:
Revert "[Refactor] convert ValidateCodeForm to TypeScript"

This reverts commit b782ad6b415d3d3f211a7005d275f1ddf4c903ab.

fedirjh committed:
Revert "[Refactor] Convert DesktopRedirectPage.js to TypeScript"

This reverts commit f7997a1ebb5da8b66c01595168b1371e125189b5.

fedirjh committed:
Revert "[Refactor] Convert DesktopRedirectPage.js to TypeScript"

This reverts commit f7997a1ebb5da8b66c01595168b1371e125189b5.

fedirjh committed:
Revert "[Refactor] Convert DesktopSignInRedirectPage to TypeScript"

This reverts commit 7a01d3d5306709b78940376a7313ef12c17eba5c.

fedirjh committed:
Revert "[Refactor] Convert DesktopSignInRedirectPage to TypeScript"

This reverts commit 7a01d3d5306709b78940376a7313ef12c17eba5c.

fedirjh committed:
Revert "[Refactor] convert to TypeScript"

This reverts commit eb862e5807d7241c3d01e795062e641345eebb9c.

fedirjh committed:
Revert "[Refactor] convert to TypeScript"

This reverts commit eb862e5807d7241c3d01e795062e641345eebb9c.

fedirjh committed:
[Chore] Fix types

fedirjh committed:
[Chore] Fix typeScript checks + Rename BaseLoginForm

fedirjh committed:
[Refactor] Migrate SignInPage to typeScript

fedirjh committed:
[Chore] Fix code style

fedirjh committed:
[Chore] Fix code style

fedirjh committed:
[Chore] Fix TS errors

fedirjh committed:
[Chore] Fix Lint & TS

fedirjh committed:
[Chore] Fix Types

fedirjh committed:
[Chore] Fix Types

s77rt commented on src/pages/signin/LoginForm/BaseLoginForm.tsx:
```diff
@@ -202,22 +165,23 @@ function LoginForm(props) {
         // Check if this login has an account associated with it or not
-        Session.beginSignIn(parsedPhoneNumber.possible ? parsedPhoneNumber.number.e164 : loginTrim);
-    }, [login, props.account, props.closeAccount, props.network, validate]);
+        Session.beginSignIn(parsedPhoneNumber.possible ? parsedPhoneNumber?.number?.e164 ?? '' : loginTrim);
```
```suggestion
        Session.beginSignIn(parsedPhoneNumber.possible && parsedPhoneNumber.number?.e164 ? parsedPhoneNumber.number.e164 : loginTrim);
```

s77rt commented on src/pages/signin/LoginForm/BaseLoginForm.tsx:
```diff
@@ -226,27 +190,30 @@ function LoginForm(props) {
+            if (!input.current) {
+                return false;
+            }
+            return input.current.isFocused() as boolean;
```
Any idea on why the type is `any`?

s77rt commented on src/pages/signin/LoginForm/BaseLoginForm.tsx:
```diff
@@ -307,23 +274,24 @@ function LoginForm(props) {
-            {!_.isEmpty(props.account.success) && <Text style={[styles.formSuccess]}>{props.account.success}</Text>}
-            {!_.isEmpty(props.closeAccount.success || props.account.message) && (
+            {account?.success && <Text style={[styles.formSuccess]}>{account.success}</Text>}
+            {(closeAccount?.success ?? account?.message) && (
```
```suggestion
            {(!!closeAccount?.success || !!account?.message) && (
```

s77rt commented on src/pages/signin/LoginForm/BaseLoginForm.tsx:
```diff
@@ -307,23 +274,24 @@ function LoginForm(props) {
                     type="success"
-                    messages={{0: props.closeAccount.success || props.account.message}}
+                    // eslint-disable-next-line @typescript-eslint/naming-convention
+                    messages={{0: (closeAccount?.success ?? account?.message ?? '') as MaybePhraseKey}}
```
Use `||` instead of `??`. Is there a reason to cast to `MaybePhraseKey`?

s77rt replied:
If `closeAccount?.success` is empty string we want to use `account?.message` as it may have an actual string.

s77rt reviewed (changes requested):
Partial review

fedirjh committed:
[Chore] Fix nullish coalescing

fedirjh committed:
[Chore] Fix nullish coalescing

fedirjh committed:
[Chore] Rename File

fedirjh committed:
[Chore] Rename File

fedirjh committed:
[Chore] Rename TS File

fedirjh committed:
[Chore] Rename TS File

fedirjh committed:
[Chore] Fix TS

s77rt commented:
@fedirjh Please do not force push as this messes up the review, breaks the `Show changes since your last review` option

s77rt commented on src/pages/signin/SignInPageLayout/Footer.tsx:
```diff
@@ -138,19 +142,21 @@ const columns = ({navigateFocus}) => [
+    const footerColumn = isVertical ? [styles.p4] : [styles.p4, isMediumScreenWidth ? styles.w50 : styles.w25];
     const footerWrapper = isVertical ? [StyleUtils.getBackgroundColorStyle(theme.signInPage), styles.overflowHidden] : [];
-
+    const getTextLinkStyle = (hovered: boolean) => [styles.footerRow, hovered ? styles.textBlue : {}];
```
```suggestion
    const getTextLinkStyle: (hovered: boolean) => StyleProp<TextStyle> = (hovered) => [styles.footerRow, hovered && styles.textBlue];
```

The signature of this function does not look correct. Either define the function type or just remove it

s77rt replied:
Yes that's the expected results. react-native accepts falsy values

https://github.com/facebook/react-native/blob/68e6a94538c3eb35de5477bc0fb6649514c1c6fc/packages/react-native/Libraries/StyleSheet/StyleSheet.d.ts#L22-L26

s77rt commented on src/pages/signin/SignInPageLayout/Footer.tsx:
```diff
@@ -164,24 +170,34 @@ function Footer(props) {
-                                                    >
-                                                        {props.translate(row.translationPath)}
-                                                    </TextLink>
+                                                    {link ? (
```
```suggestion
                                                    {!!link ? (
```

Render conditions should be `boolean` 

(do same with `onPress`)

s77rt commented on src/pages/signin/SignInPageLayout/Footer.tsx:
```diff
@@ -1,33 +1,37 @@
+    rows: FooterColumnRow[];
+};
+
+const columns = ({navigateFocus}: {navigateFocus: () => void}): FooterColumnData[] => [
```
```suggestion
const columns = ({navigateFocus}: Pick<FooterProps, 'navigateFocus'>): FooterColumnData[] => [
```

s77rt commented on src/pages/signin/SignInPageLayout/Footer.tsx:
```diff
@@ -1,33 +1,37 @@
 
-const defaultProps = {
-    shouldShowSmallScreen: false,
+type FooterColumnRow = ({onPress: () => void; link?: string} | {onPress?: () => void; link: string}) & {
```
```suggestion
type FooterColumnRow = (LinkProps | PressProps) & {
```

Import LinkProps and PressProps from TextLink

fedirjh committed:
Apply suggestions from code review

Co-authored-by: Abdelhafidh Belalia <16493223+s77rt@users.noreply.github.com>

fedirjh committed:
[Chore] Fix TS types: use TextLink types

fedirjh committed:
[Chore] Fix SAMLSignIn Props

fedirjh committed:
[Chore] Fix Lint & Address code review

fedirjh committed:
[Refactor] SignInPageLayout to use shared types

Reorganized code by introducing shared types in `SignInPageLayoutProps`, which streamlined the props of all components in the SignInPageLayout. The shared types are now used in Footer, SignInPageContent, SignInPageLayout, SignInHeroCopy, SignInPageHero, and SignInHeroImage components. Files have been moved to the SignInPageLayout directory for better organization.

s77rt commented on src/pages/signin/SignInPage.tsx:
```diff
@@ -190,13 +180,13 @@ function SignInPageInner({credentials, account, isInModal, activeClients, prefer
+        hasValidateCode: !!credentials?.validateCode,
         account,
-        isPrimaryLogin: !account.primaryLogin || account.primaryLogin === credentials.login,
+        isPrimaryLogin: !account?.primaryLogin || account?.primaryLogin === credentials?.login,
```
```suggestion
        isPrimaryLogin: !account?.primaryLogin || account.primaryLogin === credentials?.login,
```

s77rt commented on src/pages/signin/SignInPage.tsx:
```diff
@@ -208,23 +198,24 @@ function SignInPageInner({credentials, account, isInModal, activeClients, prefer
             welcomeText = isUsingRecoveryCode ? translate('validateCodeForm.enterRecoveryCode') : translate('validateCodeForm.enterAuthenticatorCode');
         } else {
-            const userLogin = Str.removeSMSDomain(credentials.login || '');
+            // eslint-disable-next-line @typescript-eslint/prefer-nullish-coalescing
+            const userLogin = Str.removeSMSDomain(credentials?.login || '');
```
```suggestion
            const userLogin = Str.removeSMSDomain(credentials?.login ?? '');
```

s77rt commented on src/pages/signin/SignInPage.tsx:
```diff
@@ -295,11 +287,13 @@ function SignInPageInner({credentials, account, isInModal, activeClients, prefer
 SignInPageInner.displayName = 'SignInPage';
 
-function SignInPage(props) {
+type SignInPageProps = SignInPageInnerProps & {navigation?: Partial<typeof navigationRef>};
```
```suggestion
type SignInPageProps = SignInPageInnerProps;
```

`navigation` prop is not being used

s77rt replied:
Correct but that prop is not being used by SigninPage. We shouldn't even pass it

s77rt replied:
I'm not sure why would the perf test but I think we should still remove that prop since it's not really used by the App. Instead add a `@ts-expect-error` comment in the perf test

fedirjh committed:
[Chore] Fix TS Types & address code review

s77rt commented on src/pages/signin/SignInPage.tsx:
```diff
@@ -154,20 +145,19 @@ function SignInPageInner({credentials, account, isInModal, activeClients, prefer
+    const isClientTheLeader = !!activeClients && ActiveClientManager.isClientTheLeader();
     // We need to show "Another login page is opened" message if the page isn't active and visible
-    // eslint-disable-next-line rulesdir/no-negated-variables
-    const shouldShowAnotherLoginPageOpenedMessage = Visibility.isVisible() && !isClientTheLeader;
```
Please revert. The problem is in the eslint rule https://github.com/Expensify/eslint-config-expensify/blob/main/eslint-plugin-expensify/no-negated-variables.js It's returning a false positive.

s77rt replied:
Yes, I don't think we should change the variable name because of a faulty eslint rule

s77rt commented on src/pages/signin/UnlinkLoginForm.tsx:
```diff
@@ -62,7 +62,7 @@ function UnlinkLoginForm({account, credentials}: UnlinkLoginFormProps) {
                 />
             )}
-            {account?.errors && !isEmptyObject(account?.errors) && (
+            {!!account?.errors && !isEmptyObject(account?.errors) && (
```
```suggestion
            {!!account?.errors && !isEmptyObject(account.errors) && (
```

s77rt commented on src/pages/signin/UnlinkLoginForm.tsx:
```diff
@@ -80,7 +80,7 @@ function UnlinkLoginForm({account, credentials}: UnlinkLoginFormProps) {
                     success
                     text={translate('unlinkLoginForm.unlink')}
-                    isLoading={!!account?.isLoading && account?.loadingForm === CONST.FORMS.UNLINK_LOGIN_FORM}
+                    isLoading={account?.isLoading && account?.loadingForm === CONST.FORMS.UNLINK_LOGIN_FORM}
```
```suggestion
                    isLoading={account?.isLoading && account.loadingForm === CONST.FORMS.UNLINK_LOGIN_FORM}
```

s77rt commented on src/pages/signin/SignInPageLayout/signInPageStyles/index.ts:
```diff
@@ -1,6 +1,6 @@
 // On web, we can use flex to fit the content to fit the viewport within a ScrollView.
-import type {ThemeStyles} from '@src/styles';
+import type ScrollViewContentContainerStyles from './types';
```
Move comment below (same for index.native.ts)

blazejkustra reviewed (approved):
Huge effort! Looks good typescript-wise 😄

fedirjh committed:
[Chore] Fix TS Types & address code review

s77rt commented:
Code looks good! Will test this asap!

s77rt commented:
## Reviewer Checklist

- [X] I have verified the author checklist is complete (all boxes are checked off).
- [X] I verified the correct issue is linked in the `### Fixed Issues` section above
- [X] I verified testing steps are clear and they cover the changes made in this PR
    - [X] I verified the steps for local testing are in the `Tests` section
    - [X] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [X] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [X] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [X] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [X] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [X] I verified tests pass on **all platforms** & I tested again on:
    - [X] Android: Native
    - [X] Android: mWeb Chrome
    - [X] iOS: Native
    - [X] iOS: mWeb Safari
    - [X] MacOS: Chrome / Safari
    - [X] MacOS: Desktop
- [X] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [X] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [X] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [X] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [X] I verified that comments were added to code that is not self explanatory
    - [X] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [X] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [X] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [X] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [X] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [X] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [X] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [X] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [X] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [X] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [X] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [X] If a new component is created I verified that:
    - [X] A similar component doesn't exist in the codebase
    - [X] All props are defined accurately and each prop has a `/** comment above it */`
    - [X] The file is named correctly
    - [X] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [X] The only data being stored in the state is data necessary for rendering and nothing else
    - [X] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [X] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [X] All JSX used for rendering exists in the render method
    - [X] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [X] If any new file was added I verified that:
    - [X] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [X] If a new CSS style is added I verified that:
    - [X] A similar style doesn't already exist
    - [X] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG`)
- [X] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [X] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [X] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [X] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [X] If the PR modifies the form input styles:
    - [X] I verified that all the inputs inside a form are aligned with each other.
    - [X] I added `Design` label so the design team can review the changes.
- [X] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [X] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [X] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

<!-- add screenshots or videos here -->

https://github.com/Expensify/App/assets/16493223/c68061fb-183e-4ac0-891c-221d087e9135


</details>

<details>
<summary>Android: mWeb Chrome</summary>

<!-- add screenshots or videos here -->

https://github.com/Expensify/App/assets/16493223/7dcb4c10-fc48-4b6d-8f40-5b8bc0edcc0a


</details>

<details>
<summary>iOS: Native</summary>

<!-- add screenshots or videos here -->

https://github.com/Expensify/App/assets/16493223/18526dd0-2174-4e94-963a-507b08ca67bd


</details>

<details>
<summary>iOS: mWeb Safari</summary>

<!-- add screenshots or videos here -->

https://github.com/Expensify/App/assets/16493223/7ded0e33-ff40-4fe6-bf1e-d12f04874104


</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

<!-- add screenshots or videos here -->

https://github.com/Expensify/App/assets/16493223/dc800d05-792a-4a15-86d2-a05f56200ec3


</details>

<details>
<summary>MacOS: Desktop</summary>

<!-- add screenshots or videos here -->
<img width="1312" alt="desktop" src="https://github.com/Expensify/App/assets/16493223/95888d89-09f3-43ed-899c-fe23b9e6070b">

</details>

s77rt commented:
cc @puneetlath

fedirjh committed:
Merge branch 'main' into ts-migration-signin

# Conflicts:
#	src/pages/signin/LoginForm/BaseLoginForm.tsx
#	src/pages/signin/SignInHeroImage.js
#	src/pages/signin/SignInPage.tsx
#	src/pages/signin/SignInPageLayout/BackgroundImage/index.ios.js
#	src/pages/signin/SignInPageLayout/Footer.tsx
#	src/pages/signin/SignInPageLayout/SignInPageContent.tsx
#	src/pages/signin/SignInPageLayout/index.tsx
#	src/types/onyx/Form.ts

fedirjh committed:
[Chore] Fix merge conflicts

fedirjh committed:
[Chore] Remove extra file

fedirjh committed:
[Chore] Remove outdated TS directives

puneetlath commented:
Good job, but man that was a big PR 😅. Any way we could break this sort of thing up more in the future?

OSBotify commented:
:hand: This PR was not deployed to staging yet because QA is ongoing. It will be automatically deployed to staging after the next production release.

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/7911420306) to staging by https://github.com/puneetlath in version: 1.4.42-0 🚀

platform | result
---|---
🤖 android 🤖|success ✅
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅
🕸 web 🕸|success ✅

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/7966217051) to production by https://github.com/chiragsalian in version: 1.4.42-5 🚀

platform | result
---|---
🤖 android 🤖|success ✅
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅
🕸 web 🕸|success ✅
